plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/NonSystemJobSuspender.java

```diff
@@ -45,41 +45,51 @@ private SuspendedJob(Job job, long scheduleDelay) {
     }
   }
 
+  private static final AtomicBoolean suspended = new AtomicBoolean(false);
+
   private static final List<SuspendedJob> suspendedJobs = new ArrayList<>();
 
   private static final NonSystemJobScheduleListener jobScheduleListener =
       new NonSystemJobScheduleListener();
 
   /** Once called, it is imperative to call {@link resume()} later. */
-  public static synchronized void suspendFutureJobs() {
-    Preconditions.checkState(!suspended, "Already suspended.");
-    suspended = true;
-    Job.getJobManager().addJobChangeListener(jobScheduleListener);
+  public static void suspendFutureJobs() {
+    synchronized (suspendedJobs) {
+      Preconditions.checkState(!suspended.getAndSet(true), "Already suspended.");
+      Job.getJobManager().addJobChangeListener(jobScheduleListener);
+    }
   }
 
-  public static synchronized void resume() {
-    Preconditions.checkState(suspended, "Not suspended.");
+  public static void resume() {
+    Preconditions.checkState(suspended.get(), "Not suspended.");
     resumeInternal();
   }
 
   @VisibleForTesting
-  static synchronized void resumeInternal() {
-    suspended = false;
-    Job.getJobManager().removeJobChangeListener(jobScheduleListener);
+  static void resumeInternal() {
+    List<SuspendedJob> copy;
+    synchronized (suspendedJobs) {
+      if (!suspended.getAndSet(false)) {
+        return;
+      }
+      Job.getJobManager().removeJobChangeListener(jobScheduleListener);
+      copy = new ArrayList<>(suspendedJobs);
+      suspendedJobs.clear();
+    }
 
-    for (SuspendedJob jobInfo : suspendedJobs) {
+    for (SuspendedJob jobInfo : copy) {
       jobInfo.job.schedule(jobInfo.scheduleDelay);
     }
-    suspendedJobs.clear();
   }
 
-  static synchronized void suspendJob(Job job, long scheduleDelay) {
-    if (suspended) {
-      if (!job.isSystem()) {
-        job.cancel();  // This will always succeed since the job is not running yet.
-        suspendedJobs.add(new SuspendedJob(job, scheduleDelay));
-      }
+  static void suspendJob(Job job, long scheduleDelay) {
+    if (!suspended.get() || job.isSystem()) {
+      return;
+    }
+    synchronized (suspendedJobs) {
+      suspendedJobs.add(new SuspendedJob(job, scheduleDelay));
     }
+    job.cancel(); // This will always succeed since the job is not running yet.
```

Comment:
(Concurrency is always very confusing to me. Spending a lot of time thinking about this.)

I'm uncertain if this is possible: what if we are about to execute `job.cancel()` (`suspended` is true at this point), but then

a) another thread was executing `resume()` (can go in because we are in the suspended state)
b) sets `suspended` to false; and
c) detaches the `JobChangeListener` (i.e., all this happening just before `job.cancel()`)?

The job will be canceled, but we end up in the normal (not suspended) state. Am I right?